jgrapht-core/src/main/java/org/jgrapht/alg/flow/min_cost/CapacityScalingMinimumCostFlow.java

```diff
@@ -0,0 +1,892 @@
+     * equal to this value is considered to be an infinite capacity arc.
+     */
+    public static final int CAP_INF = 1000 * 1000 * 1000;
+    /**
+     * A cost which is considered to be infinite. This value is used internally for flow network transformation.
+     * That is why arcs with cost magnitude greater than or equal to this value are not allowed.
+     */
+    public static final double COST_INF = 1e9;
+    /**
+     * Default scaling factor
+     */
+    public static final int DEFAULT_SCALING_FACTOR = 2;
+    /**
+     * Debug variable
+     */
+    private static final boolean DEBUG = false;
+    /**
+     * Scaling factor of this algorithm
+     */
+    private final int scalingFactor;
+    /**
+     * Variable that is used to determine whether a vertex has been labeled temporarily or permanently during
+     * Dijkstra's algorithm
+     */
+    private int counter = 1;
+    /**
+     * Specified minimum cost flow problem
+     */
+    private MinimumCostFlowProblem<V, E> problem;
+    /**
+     * Computed minimum cost flow
+     */
+    private MinimumCostFLow<E> minimumCostFLow;
+
+    private DualSolution<V, E> dualSolution;
+    /**
+     * Array of internal nodes used by the algorithm. Node: these nodes are stored in the same order as vertices of
+     * the specified flow network. This allows to determine quickly their counterparts in the network.
+     */
+    private Node[] nodes;
+    /**
+     * Array of internal arcs. Note: these arcs are stored in the same order as edges of the specified flow network.
+     * This allows to determine quickly their counterparts in the network.
+     */
+    private Arc[] arcs;
+    /**
+     * List of vertices of the flow network.
+     */
+    private List<V> graphVertices;
+    /**
+     * List of edges of the flow network.
+     */
+    private List<E> graphEdges;
+    /**
+     * Number of vertices in the network
+     */
+    private int n;
+    /**
+     * Number of edges in the network
+     */
+    private int m;
+
+    /**
+     * Constructs a new instance of the algorithm which uses default scaling factor.
+     *
+     * @param problem a minimum cost flow problem
+     */
+    public CapacityScalingMinimumCostFlow(MinimumCostFlowProblem<V, E> problem) {
+        this(problem, DEFAULT_SCALING_FACTOR);
+    }
+
+    /**
+     * Constructs a new instance of the algorithm with custom {@code scalingFactor}. If the {@code scalingFactor}
+     * is less than 2, the algorithm doesn't use scaling.
+     *
+     * @param problem       a minimum cost flow problem
+     * @param scalingFactor custom scaling factor
+     */
+    public CapacityScalingMinimumCostFlow(MinimumCostFlowProblem<V, E> problem, int scalingFactor) {
+        if (problem.graph.getType().isUndirected()) {
+            throw new IllegalArgumentException("The algorithm doesn't support undirected flow networks");
```

Comment:
It can operate on undirected networks by reducing them to directed networks, but this reduction doesn't work when the network has nonzero lower arc capacities (plus other restrictions). Should I add this reduction?

Reply:
I would not perform the reduction if it comes with side constraints such as the arc capacity or weight cannot take certain values.
It seems that the problem is still well defined if you would state the constraints (minimize costs while satisfying demands and capacities) on an undirected graph. We did something similar for some of the other flow algorithms. Hence my question whether the algorithm could function on undirected graphs as well.

Reply:
I haven't found any methods to reduce undirected flow networks with lower bounds to directed flow networks. 
Here's what I think about this problem: let's take a flow network with 2 nodes of 0 demand and n undirected edges. Let's assign some arbitrary lower bounds to these edges, which are equal to their upper bounds. Then to determine, whether this flow network has a feasible solution, we have to send an equal amount of flow to the first node and to the second. This is basically a two-partition problem., which is weakly NP-complete and consequently, we can't solve this problem in polynomial time. Therefore, the problem with the undirected flow networks with lower bounds is computationally harder than without.  

What do you think about this?